content/blog/waldo-0-3-0/index.Rmd

```diff
@@ -0,0 +1,137 @@
+---
+output: hugodown::hugo_document
+
+slug: waldo-0-3-0
+title: waldo 0.3.0
+date: 2021-08-20
+author: Hadley Wickham
+description: >
+    waldo 0.3.0 improves the display of data frames differences, and gives the 
+    objects being compared the ability to control the detail of their 
+    comparisons.
+
+photo:
+  url: https://unsplash.com/photos/E9tFH39iRPE
+  author: NordWood Themes
+
+# one of: "deep-dive", "learn", "package", "programming", or "other"
+categories: [package, programming] 
+tags: [testthat, waldo]
+---
+
+<!--
+TODO:
+* [x] Look over / edit the post's title in the yaml
+* [x] Edit (or delete) the description; note this appears in the Twitter card
+* [x] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
+* [x] Find photo & update yaml metadata
+* [x] Create `thumbnail-sq.jpg`; height and width should be equal
+* [x] Create `thumbnail-wd.jpg`; width should be >5x height
+* [x] `hugodown::use_tidy_thumbnails()`
+* [x] Add intro sentence, e.g. the standard tagline for the package
+* [x] `usethis::use_tidy_thanks()`
+-->
+
+We're delighted to announce the release of [waldo](https://waldo.r-lib.org) 0.3.0. waldo is designed to find and concisely describe the difference between a pair of R objects. It was designed primarily to improve failure messages for `testthat::expect_equal()`, but it turns out to be useful in a number of other situations.
+
+You can install it from CRAN with:
+
+```{r, eval = FALSE}
+install.packages("waldo")
+```
+
+This blog post highlights the two biggest changes in this release: a new display format for data frame differences and new tools for package developers to control the details of comparison. You can see a full list of changes in the [release notes](https://github.com/r-lib/waldo/blob/master/NEWS.md)
+
+```{r setup}
+library(waldo)
+```
+
+## Data frame differences
+
+waldo 0.2.0 treated data frames in the same way as lists, which worked fine if a column changed, but wasn't terribly informative if a row changed. In 0.3.0, data frames get a new row-oriented comparison:
+
+```{r}
+df1 <- data.frame(x = c(1, 2, 3, 4, 5), y = c("a", "b", "c", "d", "e"))
+df2 <- data.frame(x = c(1, 2, 3, 10, 4, 5), y = c("a", "b", "c", "X", "d", "e"))
+compare(df1, df2)
+```
+
+You'll notice that you still get the column comparison as well. This is important because the row-oriented comparison relies on the printed representation of the data frames, and there are important cases where data frames look the same but are actually different. The most important case is probably strings vs factors: example:
```

Comment:
```suggestion
You'll notice that you still get the column comparison as well. This is important because the row-oriented comparison relies on the printed representation of the data frames, and there are cases where data frames look the same but are actually different. The most important case is probably strings vs factors: example:
```

Reply:
Or replace "important" with another word, e.g. key, just so you don't have it twice in that sentence.